src/main/java/org/aksw/simba/lemming/creation/Inferer.java

```diff
@@ -189,6 +253,96 @@ public OntModel readOntology(String filePath) {
 
 		return ontModel;
 	}
-	
-	
+
+	/**
+	 * Searches for the equivalents in an ontology and maps them to our Equivalent<T
+	 * extends OntResource> class, producing a map of the Equivalent objects with
+	 * the URIs as keys.
+	 * 
+	 * @see Equivalent<T extends OntResource>
+	 * @param <T>
+	 * @param ontElements the ontology classes or properties
+	 * @return
+	 */
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	public <T extends OntResource> Map<String, Equivalent> searchEquivalents(Set ontElements) {
+
+		Map<String, Equivalent> uriNodeMap = new HashMap<String, Equivalent>();
+
+		Stack<T> stack = new Stack<T>();
+		stack.addAll(ontElements);
+
+		Map<String, Boolean> visitedMap = (Map<String, Boolean>) ontElements.stream()
+				.collect(Collectors.toMap(Object::toString, visited -> false));
+
+		Set<Equivalent> elements = new HashSet<Equivalent>();
+
+		while (stack.size() > 0) {
+			T current = stack.pop();
+			boolean isSame = false;
+			boolean isVisited = visitedMap.get(current.toString());
+			if (!isVisited) {
+				List eqsList = null;
+				try {
+					if (current instanceof OntProperty)
+						eqsList = ((OntProperty) current).listEquivalentProperties().toList();
+					if (current instanceof OntClass)
+						eqsList = ((OntClass) current).listEquivalentClasses().toList();
+				} catch (ConversionException e) {
+					LOGGER.warn(
+							"Cannot convert the equivalents. The ontology does not have any further info on the equivalents of {}.",
+							current.toString());
+				}
+
+				if (eqsList != null && !eqsList.isEmpty()) {
+					eqsList.forEach(property -> {
+						visitedMap.putIfAbsent(property.toString(), false);
+					});
+					stack.addAll(eqsList);
+				}
+
+				// check to which node do we need to add this info to
+				Iterator<Equivalent> propIterator = elements.iterator();
+				while (propIterator.hasNext()) {
+					Equivalent curNode = propIterator.next();
+					isSame = curNode.containsElement(current);
+					if (isSame) {
+						curNode.addEquivalent(current);
+						uriNodeMap.put(current.toString(), curNode);
+						break;
+					}
+				}
+
+				// if not, create new one
+				if (!isSame) {
+					Equivalent node = new Equivalent(current);
+					if (eqsList != null) {
+						node.addEquivalentGroup((Set) eqsList.stream().collect(Collectors.toSet()));
+					}
+					elements.add(node);
+					uriNodeMap.put(current.toString(), node);
+				}
+				visitedMap.put(current.toString(), true);
+			}
+		}
+		return uriNodeMap;

```

Comment:
Does your algorithm execute an inferencer which makes all the equivalence relationships explicit?

Reply:
Hopefully, yes! The idea was to treat all the properties/classes present in the ontology before the inference. This way, when the actual inference takes place in the RDF Model, all the domains and ranges of the equivalence relationships established, would be explicit in these "Equivalent" objects.